frame/support/procedural/src/compact_pallet_error.rs

```diff
@@ -0,0 +1,116 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use frame_support_procedural_tools::generate_crate_access_2018;
+use std::convert::identity;
+
+// Derive `CompactPalletError`
+pub fn derive_compact_pallet_error(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+	let syn::DeriveInput { ident: name, generics, data, .. } = match syn::parse(input) {
+		Ok(input) => input,
+		Err(e) => return e.to_compile_error().into(),
+	};
+
+	let frame_support = match generate_crate_access_2018("frame-support") {
+		Ok(c) => c,
+		Err(e) => return e.into_compile_error().into(),
+	};
+	let frame_support = &frame_support;
+	let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+	let compactness_check = match data {
+		syn::Data::Struct(syn::DataStruct { struct_token, fields, .. }) => {
+			if fields.len() > 1 {
+				let msg = "Cannot derive `CompactPalletError` for structs with more than 1 field";
+				return syn::Error::new(struct_token.span, msg).into_compile_error().into()
+			}
+
+			match fields {
+				syn::Fields::Named(mut f) if f.named.len() == 1 => {
+					let field_ty = f.named.pop().unwrap().into_value().ty;
+					quote::quote! {
+						<
+							#field_ty as #frame_support::traits::CompactPalletError
+						>::check_compactness()
+					}
+				},
+				syn::Fields::Unnamed(mut f) if f.unnamed.len() == 1 => {
+					let field_ty = f.unnamed.pop().unwrap().into_value().ty;
+					quote::quote! {
+						<
+							#field_ty as #frame_support::traits::CompactPalletError
+						>::check_compactness()
+					}
+				},
+				_ => quote::quote!(true),
+			}
+		},
+		syn::Data::Enum(syn::DataEnum { variants, .. }) => {
+			let field_tys = variants
+				.into_iter()
+				.map(|variant| match variant.fields {
+					syn::Fields::Named(mut f) if f.named.len() == 1 =>
+						Ok(Some(f.named.pop().unwrap().into_value().ty)),
+					syn::Fields::Unnamed(mut f) if f.unnamed.len() == 1 =>
+						Ok(Some(f.unnamed.pop().unwrap().into_value().ty)),
+					syn::Fields::Unit => Ok(None),
+					_ if variant.ident == "__Ignore" => Ok(None),
```

Comment:
it may not be very good to match on variant name here, this macro will be used by user outside, they may use the name __Ignore for their own usage.
Maybe one solution is to implement CompactPalletError for the type `(PhantomData<T>, Never)` and use this type when generating the type Error.

Reply:
But if they do, then it wouldn't actually compile, because we create an `__Ignore` variant for the error enum during macro expansion, and it would clash with their declaration.

Reply:
I don't understand what you mean, can you explain more ?
AFAICT instead of making __Ignore a special case I suggest we have a type which implement CompactPalletError and can capture generics, this type in my suggestion above is `(PhantomData<T>, Never)`. this type cannot be constructed so it is safe to say that it is a compact pallet error.
Then when deriving the CompactPalletError for the error enum the generated code will do something like:
```
<(PhantomData<T>, Never) as CompactPalletError>::check_compactness()
```
And this will return true.

Reply:
We always construct an `__Ignore` enum variant with pallet errors:

https://github.com/paritytech/substrate/blob/4de8ee2dee389f4246f3156667fc3d6df7047e4f/frame/support/procedural/src/pallet/expand/error.rs#L34-L39

https://github.com/paritytech/substrate/blob/4de8ee2dee389f4246f3156667fc3d6df7047e4f/frame/support/src/error.rs#L93-L98

Thus if the user created an `__Ignore` variant within a pallet error enum, their project will simply not compile because it clashes with our macro definition of this variant. So matching with the ident is fine, we don't necessarily need to implement the trait on the type.

Reply:
with nested error, I think user can define their own error outside the pallet macro, and use the derive macro to derive the CompactPalletError trait. In this case the __Ignore variant will not conflict.

Reply:
Okay, so I instead instructed the algorithm to check the ident of the types that are being compared and see if we're dealing with `(PhantomData<T>, Never)`; if so, then we simply skip it and don't call `check_compactness` on it. Since I now have to derive `Encode` and `Decode`, the `Never` type will also have to have an implementation of `Encode` and `Decode`, and it might not really make sense for it to have one since it is the kind of type that can never be constructed.

Reply:
maybe we can make use of skip attribute in order not to implement Encode/Decode for Never, but having Never implementing Encode/Decode is also fine (as long as decode always fail).

your solution of skipping the type `(PhantomData<T>, Never)` for checking the compactness seems equivalent as implementing CompactPalletError on `(PhantomData<T>, Never)`directly, but I'm fine with both, (though I prefer to implement CompactPalletError, because matching on names is always slightly wrong given that user can bring its own name into scope).